File: gseim_grc/src/gseim_cpp_lib/gseim_solver_test.py
import filecmp
import os
import subprocess
import sys

from importlib_resources import files
import pytest

def _test_solver(fname):
    r = subprocess.run(
        [
            str(files('gseim_cpp_lib').joinpath('gseim_solver')),
            str(files('gseim_cpp_lib').joinpath('test_data', 'input', fname)),
        ],
        capture_output=True,
        text=True,
        env={'HOME': 'not set'},
    )

    assert 'Program completed.' in r.stdout
    r.check_returncode()

    print(r.stdout)

    out_fname = os.path.splitext(fname)[0] + '.dat'

    assert filecmp.cmp(
        str(files('gseim_cpp_lib').joinpath('test_data', 'input', out_fname)),
        str(files('gseim_cpp_lib').joinpath('test_data', 'output', out_fname)),
    )

def test_ac_controllerer_1():
    _test_solver('ac_controller_1.in')

def test_buck():
    _test_solver('buck.in')

def test_controlled_rectifier_2():
    _test_solver('controlled_rectifier_2.in')

def test_test_1():
    _test_solver('test_1.in')

def test_test_2():
    _test_solver('test_2.in')

def test_test_3():
    _test_solver('test_3.in')

def test_test_4():
    _test_solver('test_4.in')

def test_test_5():
    _test_solver('test_5.in')

def test_test_6():
    _test_solver('test_6.in')

def test_test_7():
    _test_solver('test_7.in')

def test_test_8():
    _test_solver('test_8.in')

def test_test_9():
    _test_solver('test_9.in')

def test_test_10():
    _test_solver('test_10.in')

def test_test_11():
    _test_solver('test_11.in')

def test_test_12():
    _test_solver('test_12.in')

def test_test_13():
    _test_solver('test_13.in')

def test_test_14():
    _test_solver('test_14.in')

def test_test_15():
    _test_solver('test_15.in')

def test_test_16():
    _test_solver('test_16.in')

def test_test_17():
    _test_solver('test_17.in')

if __name__ == '__main__':
    sys.exit(pytest.main(sys.argv))

File: gseim_grc/src/gseim_cpp_lib/gseim_solver_test.cpp
#include <cstdio>
#include <iostream>
#include <filesystem>

#include "gseim_cpp_lib/gseim_solver.h"

using namespace std;

namespace fs = std::filesystem;

int diff(string fname1, string fname2) {
    string diff_cmd = string("diff --unified ") + fname1.c_str() + " " + fname2.c_str();
    FILE *fp = popen(diff_cmd.c_str(), "r");
    if (fp == NULL) {
        cout << "Could not run diff on files" << endl;
    }

    int buf_size = 255;
    char buf[buf_size];
    while (fgets(buf, buf_size, fp) != NULL) {
        cout << buf;
    }

    int retcode = pclose(fp);
    return retcode;
}

void test_solver(string fname) {
    fs::path input_data_path = fs::current_path()/"gseim_cpp_lib"/"test_data"/"input"/fname;

    string input_path_str = input_data_path.string();
    const char* argv[] = {"MISSING", input_path_str.c_str()};

    int retcode = solve(2, (char**)argv);
    if (retcode != 0) {
        cout << "solve for " << fname << " generated return code: " << retcode << endl;
        exit(1);
    }

    fs::path output_path = fs::current_path()/"gseim_cpp_lib"/"test_data"/"input"/fname;
    output_path.replace_extension("dat");

    fs::path expected_output_path = fs::current_path()/"gseim_cpp_lib"/"test_data"/"output"/fname;
    expected_output_path.replace_extension("dat");

    retcode = diff(output_path.string(), expected_output_path.string());
    if (retcode != 0) {
        cout << "diff failed" << endl;
        cout << "diff return code: " << retcode << endl;
        exit(1);
    }
}

int main() {
    test_solver("ac_controller_1.in");
    test_solver("buck.in");
    test_solver("controlled_rectifier_2.in");

    return 0;
}
